### PR TITLE
Master wowl aab

### PR DESCRIPTION
**Pull request by aab-odoo**

### CODE REVIEW

File: addons/wowl/__manifest__.py [hunk 1]
```diff
@@ -31,6 +31,7 @@
             'static/src/download/**/*',
             'static/src/errors/**/*',
             'static/src/debug/**/*',
+            'static/src/switch_company_menu/**/*',
```
Comment (ged-odoo) [inline addons/wowl/__manifest__.py]: in webclient

File: addons/wowl/static/src/actions/action_container.js [hunk 2]
```diff
@@ -42,19 +43,58 @@ export class ActionContainer extends Component {
...
+  onDialogClosed() {
     this.dialog = {};
     this.render();
   }
-  shouldUpdate(nextProps) {
-    // We should not be updated by a render triggered from our parent
-    // LPE FIXME: except in the case of the HomeMenu
-    return false;
+
+  /**
```
Comment (ged-odoo) [inline addons/wowl/static/src/actions/action_container.js]: service `scroller`

File: addons/wowl/static/src/actions/action_hook.js [hunk 3]
```diff
@@ -28,7 +28,7 @@ function getScrollPosition(component) {
  * Set top and left scroll positions to the given values. By default, the
  * scrolling area is the '.o_content' main div. In mobile, it is the body.
  */
-function setScrollPosition(component, offset) {
+export function setScrollPosition(component, offset) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/actions/action_hook.js]: move to scroller service

File: addons/wowl/static/src/actions/action_hook.js [hunk 4]
```diff
@@ -48,24 +48,33 @@ function setScrollPosition(component, offset) {
...
+  }
+
   onMounted(() => {
     if (component.props.state) {
       setScrollPosition(component, component.props.state[scrollSymbol]);
     }
     if (params.beforeLeave && component.props.__beforeLeave__) {
       component.props.__beforeLeave__(params.beforeLeave);
     }
+    component.env.bus.on("ACTION_MANAGER:EXPORT_CONTROLLER_STATE", component, exportState);
```
Comment (ged-odoo) [inline addons/wowl/static/src/actions/action_hook.js]: useBus ?

File: addons/wowl/static/src/actions/action_service.js [hunk 5]
```diff
@@ -193,6 +211,46 @@ function makeActionManager(env) {
...
+      index = options.index;
+    } else {
+      index = controllerStack.length;
+    }
+    return index;
+  }
+
   /**
    * Triggers a re-rendering with respect to the given controller.
    *
```
Comment (ged-odoo) [inline addons/wowl/static/src/actions/action_service.js:256]: constructor => setup

File: addons/wowl/static/src/components/dropdown/dropdown.js [hunk 6]
```diff
@@ -5,20 +5,25 @@ import { scrollTo } from "../../utils/scroll_to";
...
     this.hotkeyService = useService("hotkey");
     this.hotkeyTokens = [];
     this.state = useState({ open: this.props.startOpen, groupIsOpen: this.props.startOpen });
+
     this.ui = useService("ui");
-    // Close on outside click listener
-    useExternalListener(window, "click", this.onWindowClicked);
-    // Listen to all dropdowns state changes
-    useBus(Dropdown.bus, "state-changed", this.onDropdownStateChanged);
+    if (!this.props.manualOnly) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/components/dropdown/dropdown.js:17]: find someone to write a test

File: addons/wowl/static/src/services/menu_service.js [hunk 7]
```diff
@@ -45,15 +52,21 @@ function makeMenus(env, menusData) {
...
-      env.services.router.pushState({
-        menu_id: `${menu.id}`,
-      });
     },
     setCurrentMenu(menu) {
       menu = typeof menu === "number" ? this.getMenu(menu) : menu;
       if (menu && menu.appID !== currentAppId) {
         currentAppId = menu.appID;
         env.bus.trigger("MENUS:APP-CHANGED");
+        env.services.router.pushState({
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/menu_service.js:63]: `pushState({menu_id: ...}, { lock: true});` ?

Comment (ged-odoo) [reply addons/wowl/static/src/services/menu_service.js:63]:
```js
pushState({menu_id: { value: ..., lock: true}}); ?
```

File: addons/wowl/static/src/services/router_service.js [hunk 8]
```diff
@@ -56,6 +55,39 @@ function getRoute() {
   return { pathname, search: searchQuery, hash: hashQuery };
 }
 
+export function makePreProcessQuery(getCurrent) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/router_service.js:62]: at least minimal docstring

File: addons/wowl/static/tests/services/router_service_tests.js [hunk 9]
```diff
@@ -113,3 +113,8 @@ QUnit.test("can redirect an URL", async (assert) => {
   assert.verifySteps(["1000", "250", "/my/test/url/2"]);
   unpatch(browser, "test.location");
 });
+
+// TODO: test
+//  - push/replaceState requests occuring in the same tick are aggregated
+//  - we don't push twice the same url (different key orders)
+//  - undefined keys are not pushed in the url
```
Comment (ged-odoo) [inline addons/wowl/static/tests/services/router_service_tests.js:120]: locked stuff

File: addons/wowl/static/src/services/user_service.js [hunk 10]
```diff
@@ -23,9 +19,29 @@ function computeAllowedCompanyIds(env) {
   return allowedCompanies;
 }
 
+export function makeSetCompanies(getAllowedCompanyIds) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/user_service.js:29]: minimal docstring

File: addons/wowl/static/src/services/user_service.js [hunk 11]
```diff
@@ -35,19 +51,34 @@ export const userService = {
...
     };
-    const cids = allowedCompanies.join(",");
-    env.services.router.replaceState({ cids });
-    env.services.cookie.setCookie("cids", cids);
+
+    cids = allowedCompanies.join(",");
+    router.replaceState({ "lock cids": cids });
+    cookie.setCookie("cids", cids);
+
+    if (Object.keys(user_companies.allowed_companies).length > 1) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/user_service.js]: use `isDisplayed` or `isVisible`

File: addons/wowl/static/src/switch_company_menu/switch_company_menu.js [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+import { useService } from "../core/hooks";
+
+export class SwitchCompanyMenu extends owl.Component {
+  static isDisplayed(env) {
+    return !env.services.ui.isSmall;
+  }
+
+  setup() {
+    this.ui = useService("ui");
```
Comment (ged-odoo) [inline addons/wowl/static/src/switch_company_menu/switch_company_menu.js]: remove

File: addons/wowl/static/src/switch_company_menu/switch_company_menu.xml [hunk 13]
```diff
@@ -0,0 +1,41 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates xml:space="preserve">
+
+<li t-name="wowl.SwitchCompanyMenu" class="o_switch_company_menu">
+    <Dropdown>
+        <span t-attf-class="{{ui.isSmall ? 'fa fa-building-o' : 'oe_topbar_name'}}">
```
Comment (ged-odoo) [inline addons/wowl/static/src/switch_company_menu/switch_company_menu.xml]: ui.issmall ? env.isSmall

File: addons/wowl/static/src/utils/concurrency.js [hunk 14]
```diff
@@ -33,3 +33,88 @@ export class KeepLast {
...
+  }
+  /**
+   * Add a computation to the queue, it will be executed as soon as the
+   * previous computations are completed.
+   *
+   * @param {function} action a function which may return a Promise
+   * @returns {Promise}
+   */
+  exec(action) {
+    var self = this;
```
Comment (ged-odoo) [inline addons/wowl/static/src/utils/concurrency.js]: es6ifier

File: addons/wowl/static/src/utils/concurrency.js [hunk 15]
```diff
@@ -33,3 +33,88 @@ export class KeepLast {
...
+    return this.lock.then(function () {
+      return result;
+    });
+  }
+  /**
+   * @returns {Promise} resolved as soon as the Mutex is unlocked
+   *   (directly if it is currently idle)
+   */
+  getUnlockedDef() {
+    return this.unlockedProm || Promise.resolve();
```
Comment (ged-odoo) [inline addons/wowl/static/src/utils/concurrency.js:118]: readd tests

File: addons/wowl/static/src/utils/misc.js [hunk 16]
```diff
@@ -100,3 +100,37 @@ export function json_node_to_xml(node, humanReadable, indent) {
     return r + "/>";
   }
 }
+
+export function objectToUrlEncodedString(obj) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/utils/misc.js]: move this into router

Comment (ged-odoo) [reply addons/wowl/static/src/utils/misc.js]: or at least in utils/url

File: addons/wowl/static/src/utils/misc.js [hunk 17]
```diff
@@ -100,3 +100,37 @@ export function json_node_to_xml(node, humanReadable, indent) {
...
+  return Object.entries(obj)
+    .map(([k, v]) => (v ? `${k}=${encodeURIComponent(v)}` : k))
+    .join("&");
+}
+
+export function urlBuilder(origin) {
```
Comment (ged-odoo) [inline addons/wowl/static/src/utils/misc.js]: add some docstring

File: addons/wowl/static/src/utils/objects.js [hunk 18]
```diff
@@ -0,0 +1,11 @@
+/** @odoo-module **/
+
+/**
+ * Shallow compares two objects.
+ */
+export function shallowEqual(obj1, obj2) {
+  return (
+    Object.keys(obj1).length === Object.keys(obj2).length &&
```
Comment (ged-odoo) [inline addons/wowl/static/src/utils/objects.js]: 2 times object.keys

Comment (ged-odoo) [reply addons/wowl/static/src/utils/objects.js]: add some tests

File: addons/wowl/static/src/webclient/user_menu/user_menu.js [hunk 19]
```diff
@@ -7,6 +7,10 @@ import { browser } from "../../core/browser";
 const { Component } = owl;
 
 export class UserMenu extends Component {
+  static isDisplayed(env) {
+    return !env.services.ui.isSmall;
```
Comment (ged-odoo) [inline addons/wowl/static/src/webclient/user_menu/user_menu.js]: env.isSmall

File: addons/wowl/static/tests/helpers/utility.js [hunk 20]
```diff
@@ -167,7 +167,7 @@ export function click(el, selector) {
     }
     target = els[0];
   }
-  const ev = new MouseEvent("click", { bubbles: true });
+  const ev = new MouseEvent("click", { bubbles: true, cancelable: true });
```
Comment (ged-odoo) [inline addons/wowl/static/tests/helpers/utility.js:173]: to check

File: addons/wowl/static/tests/webclient/switch_company_menu_tests.js [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+  }
+  Parent.components = { SwitchCompanyMenu };
+  Parent.template = owl.tags.xml`
+    <div>
+      <SwitchCompanyMenu/>
+    </div>`;
+  async function createParent() {
+    const env = await makeTestEnv(testConfig);
+    const target = getFixture();
+    return mount(Parent, { env, target });
```
Comment (ged-odoo) [inline addons/wowl/static/tests/webclient/switch_company_menu_tests.js]: check if can remove parent

File: addons/wowl/static/tests/webclient/switch_company_menu_tests.js [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+import { click } from "../helpers/utility";
+import { SwitchCompanyMenu } from "../../src/switch_company_menu/switch_company_menu";
+import { hotkeyService } from "../../src/services/hotkey_service";
+
+const { mount } = owl;
+
+QUnit.module("SwitchCompanyMenu", (hooks) => {
+  let testConfig;
+  class Parent extends owl.Component {
+    get systrayItems() {
```
Comment (ged-odoo) [inline addons/wowl/static/tests/webclient/switch_company_menu_tests.js]: try to remove

File: addons/wowl/static/tests/webclient/switch_company_menu_tests.js [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+    serviceRegistry.add("user", makeFakeUserService({ session_info }));
+    serviceRegistry.add("hotkey", hotkeyService);
+    testConfig = { serviceRegistry };
+  });
+
+  QUnit.test("renders", async (assert) => {
+    assert.expect(9);
+    const parent = await createParent();
+    assert.containsOnce(parent, "li.o_switch_company_menu");
+    assert.strictEqual(parent.el.querySelector(".o_switch_company_menu").textContent, "Hermit");
```
Comment (ged-odoo) [inline addons/wowl/static/tests/webclient/switch_company_menu_tests.js]: assert.containsText(..., "Hermit")

File: addons/wowl/static/src/services/menu_service.js (hunk 7 above)
Comment (aab-odoo) [reply addons/wowl/static/src/services/menu_service.js:63]: I wrote a FIXME, but I didn't change anything yet

File: addons/wowl/static/src/services/scroller_service.js [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+      } else {
+        scrollingEl = component.el.querySelector(".o_action_manager .o_content");
+      }
+      if (scrollingEl) {
+        scrollingEl.scrollLeft = offset.left || 0;
+        scrollingEl.scrollTop = offset.top || 0;
+      }
+    }
+
+    await whenReady();
```
Comment (ged-odoo) [inline addons/wowl/static/src/services/scroller_service.js]: sure it is needed?

**Pull request merged**